plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/ptf/Sheet.java

```diff
@@ -0,0 +1,152 @@
+    {
+        this.metadata = new SheetsMetadata(requireNonNull(client, "client is null"));
+    }
+
+    @Override
+    public ConnectorTableFunction get()
+    {
+        return new SheetFunction(metadata);
+    }
+
+    public static class SheetFunction
+            extends AbstractConnectorTableFunction
+    {
+        private final SheetsMetadata metadata;
+
+        public SheetFunction(SheetsMetadata metadata)
+        {
+            super(
+                    SCHEMA_NAME,
+                    NAME,
+                    ImmutableList.of(
+                            ScalarArgumentSpecification.builder()
+                                    .name(ID_ARGUMENT)
+                                    .type(VARCHAR)
+                                    .build(),
+                            ScalarArgumentSpecification.builder()
+                                    .name(RANGE_ARGUMENT)
+                                    .type(VARCHAR)
+                                    .defaultValue(utf8Slice(DEFAULT_RANGE))
+                                    .build()),
+                    GENERIC_TABLE);
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments)
+        {
+            String sheetId = ((Slice) ((ScalarArgument) arguments.get(ID_ARGUMENT)).getValue()).toStringUtf8();
+            validateSheetId(sheetId);
+            String rangeArgument = ((Slice) ((ScalarArgument) arguments.get(RANGE_ARGUMENT)).getValue()).toStringUtf8();
+
+            SheetsConnectorTableHandle tableHandle = new SheetsSheetTableHandle(sheetId, rangeArgument);
+            SheetFunctionHandle handle = new SheetFunctionHandle(tableHandle);
+
+            List<Descriptor.Field> fields = metadata.getColumnHandles(session, tableHandle).entrySet().stream()
+                    .map(entry -> new Descriptor.Field(
+                            entry.getKey(),
+                            Optional.of(((SheetsColumnHandle) (entry.getValue())).getColumnType())))
+                    .collect(toImmutableList());
+
+            Descriptor returnedType = new Descriptor(fields);
+
+            return TableFunctionAnalysis.builder()
+                    .returnedType(returnedType)
+                    .handle(handle)
+                    .build();
+        }
+    }
+
+    private static void validateSheetId(String sheetId)
+    {
+        // Sheet ids cannot contain "#", if a "#" is present it is because a range is present
+        // Ranges should be provided through the range argument
+        // https://developers.google.com/sheets/api/guides/concepts
+        if (sheetId.contains(RANGE_SEPARATOR)) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Google sheet ID %s cannot contain '#'. Provide a range through the 'range' argument.".formatted(sheetId));
```

Comment:
`"Google sheet ID %s cannot contain '#'.
`
is this already documented?

Reply:
See the link: https://developers.google.com/sheets/api/guides/concepts Google sheet ids by definition cannot contain '#'. 

Trino uses '#' as a separator character in the metadata sheet mapping so people can query tabs and ranges within a sheet. For example:) `sheet_id#tab_name!D12:E13`

This error message is because the sheet function has separated named parameters for the id and range respectively while the metadata sheet uses a single field with a '#' separator between the id and range. 

If someone puts a '#' in the id field, it is an error.